src/schedule/zephyr_ll.c

```diff
@@ -198,44 +199,51 @@ static void zephyr_ll_run(void *data)
 			continue;
 		}
 
-		pdata->run = true;
-		task->state = SOF_TASK_STATE_RUNNING;
+		if (domain_is_pending(domain, task, NULL)) {
+			pdata->run = true;
+			task->state = SOF_TASK_STATE_RUNNING;
+		}
 
 		/* Move the task to a temporary list */
 		list_item_del(list);
 		list_item_append(list, &task_head);
 
-		zephyr_ll_unlock(sch, &flags);
-
-		/*
-		 * task's .run() should only return either
-		 * SOF_TASK_STATE_COMPLETED or SOF_TASK_STATE_RESCHEDULE
-		 */
-		state = do_task_run(task);
-		if (state != SOF_TASK_STATE_COMPLETED &&
-		    state != SOF_TASK_STATE_RESCHEDULE) {
-			tr_err(&ll_tr,
-			       "zephyr_ll_run: invalid return state %u",
-			       state);
-			state = SOF_TASK_STATE_RESCHEDULE;
-		}
-
-		zephyr_ll_lock(sch, &flags);
+		if (task->state == SOF_TASK_STATE_RUNNING) {
+			zephyr_ll_unlock(sch, &flags);
 
-		if (pdata->freeing || state == SOF_TASK_STATE_COMPLETED) {
-			zephyr_ll_task_done(sch, task);
-		} else {
 			/*
-			 * task->state could've been changed to
-			 * SOF_TASK_STATE_CANCEL
+			 * task's .run() should only return either
+			 * SOF_TASK_STATE_COMPLETED or SOF_TASK_STATE_RESCHEDULE
 			 */
-			switch (task->state) {
-			case SOF_TASK_STATE_CANCEL:
+
+			state = do_task_run(task);
+			if (state != SOF_TASK_STATE_COMPLETED &&
+			    state != SOF_TASK_STATE_RESCHEDULE) {
+				tr_err(&ll_tr,
+				       "zephyr_ll_run: invalid return state %u",
+				       state);
+				state = SOF_TASK_STATE_RESCHEDULE;
+			}
+
+			zephyr_ll_lock(sch, &flags);
+
+			if (pdata->freeing || state == SOF_TASK_STATE_COMPLETED) {
 				zephyr_ll_task_done(sch, task);
-				break;
-			default:
-				break;
+			} else {
+				/*
+				 * task->state could've been changed to
+				 * SOF_TASK_STATE_CANCEL
+				 */
+				switch (task->state) {
+				case SOF_TASK_STATE_CANCEL:
+					zephyr_ll_task_done(sch, task);
+					break;
+				default:
+					break;
+				}
 			}
+			/* update task state */
+			task->state = state;
```

Comment:
There are a few changes in this commit, including this one, that modify behaviour of the (pipeline) task scheduler which is very central to SOF. If anything this will need a thorough review and extensive testing.

Reply:
...and we have a failure in https://sof-ci.01.org/sofpr/PR8653/build1331/devicetest/index.html?model=MTLP_RVP_NOCODEC&testcase=multiple-pipeline-playback

Reply:
and multiple failures in https://sof-ci.01.org/sofpr/PR8653/build1332/devicetest/index.html

Reply:
Yes, we need a lot of testing about this change.
About the failures, I can't get any info about what make the CI fail, can you please clarify it to me?